File: mcdowell/python/ch2-04.py
from ch2_linked_list import LinkedList


def partitionList(l, x):
    if l.first is None or l.first.next is None:
        return

    node = l.first

    while node.next is not None:
        if node.next.val < x:
            oldNextNext = node.next.next
            node.next.next = l.first
            l.first = node.next
            node.next = oldNextNext
        else:
            node = node.next


def test():
    l = LinkedList()
    l.addValue(1)
    l.addValue(2)
    l.addValue(3)
    l.addValue(32)
    l.addValue(30)
    l.addValue(3)
    l.addValue(4)
    l.addValue(5)
    partitionList(l, 8)

    print l.getAll()



if __name__ == '__main__':
    test()

File: mcdowell/php/ch2-04.php
<?php
require "ch2-linked-list.php";


function partitionList(LinkedList $l, $x) {

    if (!$l->first) return;
    $node = $l->first;


    while ($node->next != null) {

        if ($node->next->val < $x) {
            $oldNextNext = $node->next->next;
            $node->next->next = $l->first;
            $l->first = $node->next;
            $node->next = $oldNextNext;
        }
        else {
            $node = $node->next;
        }

    }

}



$l = new LinkedList();
$l->addValue(2);
$l->addValue(3);
$l->addValue(5);
$l->addValue(9);
$l->addValue(20);
$l->addValue(5);
$l->addValue(5);
$l->addValue(25);

print_r($l->getAll());
partitionList($l, 8);
print_r($l->getAll());